In [2]:
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [3]:
%sql postgresql://postgres:postgres@localhost:5432/analysis

'Connected: postgres@analysis'

# Extracting Information by Grouping and Summarizing

### Creating the 2014 Library Data Table

In [4]:
%%sql

CREATE TABLE pls_fy2014_pupld14a (
    stabr varchar(2) NOT NULL,
    fscskey varchar(6) CONSTRAINT fscskey2014_key PRIMARY KEY,
    libid varchar(20) NOT NULL,
    libname varchar(100) NOT NULL,
    obereg varchar(2) NOT NULL,
    rstatus integer NOT NULL,
    statstru varchar(2) NOT NULL,
    statname varchar(2) NOT NULL,
    stataddr varchar(2) NOT NULL,
    longitud numeric(10,7) NOT NULL,
    latitude numeric(10,7) NOT NULL,
    fipsst varchar(2) NOT NULL,
    fipsco varchar(3) NOT NULL,
    address varchar(35) NOT NULL,
    city varchar(20) NOT NULL,
    zip varchar(5) NOT NULL,
    zip4 varchar(4) NOT NULL,
    cnty varchar(20) NOT NULL,
    phone varchar(10) NOT NULL,
    c_relatn varchar(2) NOT NULL,
    c_legbas varchar(2) NOT NULL,
    c_admin varchar(2) NOT NULL,
    geocode varchar(3) NOT NULL,
    lsabound varchar(1) NOT NULL,
    startdat varchar(10),
    enddate varchar(10),
    popu_lsa integer NOT NULL,
    centlib integer NOT NULL,
    branlib integer NOT NULL,
    bkmob integer NOT NULL,
    master numeric(8,2) NOT NULL,
    libraria numeric(8,2) NOT NULL,
    totstaff numeric(8,2) NOT NULL,
    locgvt integer NOT NULL,
    stgvt integer NOT NULL,
    fedgvt integer NOT NULL,
    totincm integer NOT NULL,
    salaries integer,
    benefit integer,
    staffexp integer,
    prmatexp integer NOT NULL,
    elmatexp integer NOT NULL,
    totexpco integer NOT NULL,
    totopexp integer NOT NULL,
    lcap_rev integer NOT NULL,
    scap_rev integer NOT NULL,
    fcap_rev integer NOT NULL,
    cap_rev integer NOT NULL,
    capital integer NOT NULL,
    bkvol integer NOT NULL,
    ebook integer NOT NULL,
    audio_ph integer NOT NULL,
    audio_dl float NOT NULL,
    video_ph integer NOT NULL,
    video_dl float NOT NULL,
    databases integer NOT NULL,
    subscrip integer NOT NULL,
    hrs_open integer NOT NULL,
    visits integer NOT NULL,
    referenc integer NOT NULL,
    regbor integer NOT NULL,
    totcir integer NOT NULL,
    kidcircl integer NOT NULL,
    elmatcir integer NOT NULL,
    loanto integer NOT NULL,
    loanfm integer NOT NULL,
    totpro integer NOT NULL,
    totatten integer NOT NULL,
    gpterms integer NOT NULL,
    pitusr integer NOT NULL,
    wifisess integer NOT NULL,
    yr_sub integer NOT NULL
);

 * postgresql://postgres:***@localhost:5432/analysis
Done.


[]

* first we create a table with the required columns
* the primary key fscskey is unique, it can serve as natural primary key
* each column is NOT NULL

In [5]:
%%sql 

CREATE INDEX libname2014_idx ON pls_fy2014_pupld14a (libname);
CREATE INDEX stabr2014_idx ON pls_fy2014_pupld14a (stabr);
CREATE INDEX city2014_idx ON pls_fy2014_pupld14a (city);
CREATE INDEX visits2014_idx ON pls_fy2014_pupld14a (visits);

 * postgresql://postgres:***@localhost:5432/analysis
Done.
Done.
Done.
Done.


[]

* we create indexes to columns we'll use for queries

In [9]:
%%sql

COPY pls_fy2014_pupld14a
FROM '/Users/ugurtigu/Documents/Learn/Docs/SQL/pls_fy2014_pupld14a.csv'
WITH (FORMAT CSV, HEADER);

 * postgresql://postgres:***@localhost:5432/analysis
9305 rows affected.


[]

* we import our data from a CSV file 

### Creating the 2009 Library Data Table

In [10]:
%%sql

CREATE TABLE pls_fy2009_pupld09a (
    stabr varchar(2) NOT NULL,
    fscskey varchar(6) CONSTRAINT fscskey2009_key PRIMARY KEY,
    libid varchar(20) NOT NULL,
    libname varchar(100) NOT NULL,
    address varchar(35) NOT NULL,
    city varchar(20) NOT NULL,
    zip varchar(5) NOT NULL,
    zip4 varchar(4) NOT NULL,
    cnty varchar(20) NOT NULL,
    phone varchar(10) NOT NULL,
    c_relatn varchar(2) NOT NULL,
    c_legbas varchar(2) NOT NULL,
    c_admin varchar(2) NOT NULL,
    geocode varchar(3) NOT NULL,
    lsabound varchar(1) NOT NULL,
    startdat varchar(10),
    enddate varchar(10),
    popu_lsa integer NOT NULL,
    centlib integer NOT NULL,
    branlib integer NOT NULL,
    bkmob integer NOT NULL,
    master numeric(8,2) NOT NULL,
    libraria numeric(8,2) NOT NULL,
    totstaff numeric(8,2) NOT NULL,
    locgvt integer NOT NULL,
    stgvt integer NOT NULL,
    fedgvt integer NOT NULL,
    totincm integer NOT NULL,
    salaries integer,
    benefit integer,
    staffexp integer,
    prmatexp integer NOT NULL,
    elmatexp integer NOT NULL,
    totexpco integer NOT NULL,
    totopexp integer NOT NULL,
    lcap_rev integer NOT NULL,
    scap_rev integer NOT NULL,
    fcap_rev integer NOT NULL,
    cap_rev integer NOT NULL,
    capital integer NOT NULL,
    bkvol integer NOT NULL,
    ebook integer NOT NULL,
    audio integer NOT NULL,
    video integer NOT NULL,
    databases integer NOT NULL,
    subscrip integer NOT NULL,
    hrs_open integer NOT NULL,
    visits integer NOT NULL,
    referenc integer NOT NULL,
    regbor integer NOT NULL,
    totcir integer NOT NULL,
    kidcircl integer NOT NULL,
    loanto integer NOT NULL,
    loanfm integer NOT NULL,
    totpro integer NOT NULL,
    totatten integer NOT NULL,
    gpterms integer NOT NULL,
    pitusr integer NOT NULL,
    yr_sub integer NOT NULL,
    obereg varchar(2) NOT NULL,
    rstatus integer NOT NULL,
    statstru varchar(2) NOT NULL,
    statname varchar(2) NOT NULL,
    stataddr varchar(2) NOT NULL,
    longitud numeric(10,7) NOT NULL,
    latitude numeric(10,7) NOT NULL,
    fipsst varchar(2) NOT NULL,
    fipsco varchar(3) NOT NULL
);

 * postgresql://postgres:***@localhost:5432/analysis
Done.


[]

In [11]:
%%sql

CREATE INDEX libname2009_idx ON pls_fy2009_pupld09a (libname);
CREATE INDEX stabr2009_idx ON pls_fy2009_pupld09a (stabr);
CREATE INDEX city2009_idx ON pls_fy2009_pupld09a (city);
CREATE INDEX visits2009_idx ON pls_fy2009_pupld09a (visits);

 * postgresql://postgres:***@localhost:5432/analysis
Done.
Done.
Done.
Done.


[]

In [ ]:
%%sql

COPY pls_fy2009_pupld09a
FROM '/Users/ugurtigu/Documents/Learn/Docs/SQL/pls_fy2009_pupld09a.csv'
WITH (FORMAT CSV, HEADER);

## Exploring the Library Data using Aggregate Functions

* Aggregate functions combine values from multiple rows and return a single resilt based on an operation on those values
    * for example the avg() function is a aggregate function

### Counting Rows and Values Using count()

In [14]:
%%sql 

SELECT count(*)
FROM pls_fy2014_pupld14a;

 * postgresql://postgres:***@localhost:5432/analysis
1 rows affected.


[(9305,)]

In [15]:
%%sql 

SELECT count(*)
FROM pls_fy2009_pupld09a;

 * postgresql://postgres:***@localhost:5432/analysis
1 rows affected.


[(9299,)]

* after importing a dataset the first step is to count the table to make sure everything expected is imported (check with the documentation of the dataset)
* you simply check the number of the rows
* the asterisk just is there, because we want to count also NOT NULL rows

#### Counting Values Present in a Column

In [16]:
%%sql

SELECT count(salaries)
FROM pls_fy2014_pupld14a;

 * postgresql://postgres:***@localhost:5432/analysis
1 rows affected.


[(5983,)]

* to return the number of rows in a specific column
* the value is less then original rows
* this is because they are so many NULL values for salaries (most libraries don't give that data)

#### Counting Distinct Values in Column

In [17]:
%%sql

SELECT count(libname)
FROM pls_fy2014_pupld14a;

 * postgresql://postgres:***@localhost:5432/analysis
1 rows affected.


[(9305,)]

In [18]:
%%sql

SELECT count(DISTINCT libname)
FROM pls_fy2014_pupld14a;

 * postgresql://postgres:***@localhost:5432/analysis
1 rows affected.


[(8515,)]

* the **DISTINCT** keyword return just unique or combinations of unique values from the column
* if we pass that to the count() function we get the number of unique values from the column we selected
* we can see that 530 library agencies return no unique values (which is because they have same names)

### Finding Maximum and Minimum Values Using max() and min()

In [20]:
%%sql

SELECT max(visits), min(visits)
FROM pls_fy2014_pupld14a;

 * postgresql://postgres:***@localhost:5432/analysis
1 rows affected.


[(17729020, -3)]

### Aggregating Data Using GROUP BY

In [21]:
%%sql

SELECT stabr
FROM pls_fy2014_pupld14a
GROUP BY stabr
ORDER BY stabr;

 * postgresql://postgres:***@localhost:5432/analysis
56 rows affected.


[('AK',),
 ('AL',),
 ('AR',),
 ('AS',),
 ('AZ',),
 ('CA',),
 ('CO',),
 ('CT',),
 ('DC',),
 ('DE',),
 ('FL',),
 ('GA',),
 ('GU',),
 ('HI',),
 ('IA',),
 ('ID',),
 ('IL',),
 ('IN',),
 ('KS',),
 ('KY',),
 ('LA',),
 ('MA',),
 ('MD',),
 ('ME',),
 ('MI',),
 ('MN',),
 ('MO',),
 ('MP',),
 ('MS',),
 ('MT',),
 ('NC',),
 ('ND',),
 ('NE',),
 ('NH',),
 ('NJ',),
 ('NM',),
 ('NV',),
 ('NY',),
 ('OH',),
 ('OK',),
 ('OR',),
 ('PA',),
 ('PR',),
 ('RI',),
 ('SC',),
 ('SD',),
 ('TN',),
 ('TX',),
 ('UT',),
 ('VA',),
 ('VI',),
 ('VT',),
 ('WA',),
 ('WI',),
 ('WV',),
 ('WY',)]

* you can group results with the **GROUP BY** according the values in one or more columns
* this allows us to make sum() or count() for every state in our table

In [23]:
%%sql

SELECT city, stabr
FROM pls_fy2014_pupld14a
GROUP BY city, stabr
ORDER BY city, stabr
LIMIT 25;

 * postgresql://postgres:***@localhost:5432/analysis
25 rows affected.


[('ABBEVILLE', 'AL'),
 ('ABBEVILLE', 'LA'),
 ('ABBEVILLE', 'SC'),
 ('ABBOTSFORD', 'WI'),
 ('ABERDEEN', 'ID'),
 ('ABERDEEN', 'SD'),
 ('ABERNATHY', 'TX'),
 ('ABILENE', 'KS'),
 ('ABILENE', 'TX'),
 ('ABINGDON', 'IL'),
 ('ABINGDON', 'VA'),
 ('ABINGTON', 'CT'),
 ('ABINGTON', 'MA'),
 ('ABINGTON', 'PA'),
 ('ABIQUIU', 'NM'),
 ('ABSECON', 'NJ'),
 ('ACCOMAC', 'VA'),
 ('ACKLEY', 'IA'),
 ('ACTON', 'MA'),
 ('ACTON', 'ME'),
 ('ACUSHNET', 'MA'),
 ('ACWORTH', 'NH'),
 ('ADA', 'OH'),
 ('ADA', 'OK'),
 ('ADAIR', 'IA')]

* you can also use more than one columns

#### Combining GROUP BY with count()

In [24]:
%%sql

SELECT stabr, count(*)
FROM pls_fy2014_pupld14a
GROUP BY stabr
ORDER BY count(*) DESC;

 * postgresql://postgres:***@localhost:5432/analysis
56 rows affected.


[('NY', 756),
 ('IL', 625),
 ('TX', 556),
 ('IA', 543),
 ('PA', 455),
 ('MI', 389),
 ('WI', 381),
 ('MA', 370),
 ('KS', 328),
 ('NJ', 296),
 ('NE', 267),
 ('ME', 266),
 ('OH', 251),
 ('IN', 237),
 ('NH', 230),
 ('AL', 224),
 ('CT', 192),
 ('TN', 186),
 ('CA', 184),
 ('VT', 182),
 ('MO', 151),
 ('MN', 137),
 ('OR', 130),
 ('KY', 119),
 ('OK', 117),
 ('CO', 114),
 ('SD', 112),
 ('ID', 103),
 ('WV', 97),
 ('NM', 95),
 ('VA', 91),
 ('AZ', 91),
 ('AK', 85),
 ('MT', 82),
 ('FL', 82),
 ('ND', 82),
 ('NC', 80),
 ('UT', 73),
 ('LA', 68),
 ('GA', 63),
 ('WA', 62),
 ('PR', 58),
 ('AR', 58),
 ('MS', 52),
 ('RI', 48),
 ('SC', 42),
 ('MD', 24),
 ('WY', 23),
 ('DE', 21),
 ('NV', 21),
 ('GU', 1),
 ('DC', 1),
 ('MP', 1),
 ('HI', 1),
 ('VI', 1),
 ('AS', 1)]

#### Using Group BY on Multiple Columns with count()

In [26]:
%%sql

SELECT stabr, stataddr, count(*)
FROM pls_fy2014_pupld14a
GROUP BY stabr, stataddr
ORDER BY stabr ASC, count(*) DESC
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/analysis
10 rows affected.


[('AK', '00', 70),
 ('AK', '15', 10),
 ('AK', '07', 5),
 ('AL', '00', 221),
 ('AL', '07', 3),
 ('AR', '00', 58),
 ('AS', '00', 1),
 ('AZ', '00', 91),
 ('CA', '00', 182),
 ('CA', '07', 2)]

* for **stataddr**:
    * **00** NO change from last year
    * **07** Moved to a nes location
    * **15** Minor adress change

#### Revisiting sum() to Examine Library Visits - with joined Tables

In [27]:
%%sql

SELECT sum(visits) AS visits_2014
FROM pls_fy2014_pupld14a
WHERE visits >= 0;

 * postgresql://postgres:***@localhost:5432/analysis
1 rows affected.


[(1425930900,)]

In [28]:
%%sql

SELECT sum(visits) AS visits_2009
FROM pls_fy2009_pupld09a
WHERE visits >= 0;

 * postgresql://postgres:***@localhost:5432/analysis
1 rows affected.


[(1591799201,)]

* we first filter the **-3** and **-1** values with the WHERE statement
* and store this result simply in the alias
* this queries sum overall visits
* we know that we have different libraries
    * 9305 in 2014
    * 9299 in 2009

In [31]:
%%sql

SELECT sum(pls14.visits) AS visits_2014,
    sum(pls09.visits) AS visits_2009
FROM pls_fy2014_pupld14a AS pls14 JOIN pls_fy2009_pupld09a AS pls09
ON pls14.fscskey = pls09.fscskey
WHERE pls14.visits >= 0 AND pls09.visits >= 0;

 * postgresql://postgres:***@localhost:5432/analysis
1 rows affected.


[(1417299241, 1585455205)]

* at the top we use the sum() aggreagte to sum the visits from each year
* when we join tables on tables' primary keys, we're declaring the aliases
* standart join (INNER JOIN) is used, the query will result only rows where the primary key values of both tbles match
* using the where clause we eleminate the minus issue
* the reason for this query is the join
* we just considered matching primary keys, agencies in both tables

#### Grouping Visit Sums by State

In [32]:
%%sql

SELECT pls14.stabr,
       sum(pls14.visits) AS visits_2014,
       sum(pls09.visits) AS visits_2009,
       round( (CAST(sum(pls14.visits) AS decimal(10,1)) - sum(pls09.visits)) /
                    sum(pls09.visits) * 100, 2 ) AS pct_change
FROM pls_fy2014_pupld14a pls14 JOIN pls_fy2009_pupld09a pls09
ON pls14.fscskey = pls09.fscskey
WHERE pls14.visits >= 0 AND pls09.visits >= 0
GROUP BY pls14.stabr
ORDER BY pct_change DESC;

 * postgresql://postgres:***@localhost:5432/analysis
53 rows affected.


[('GU', 103593, 60763, Decimal('70.49')),
 ('DC', 4230790, 2944774, Decimal('43.67')),
 ('LA', 17242110, 15591805, Decimal('10.58')),
 ('MT', 4582604, 4386504, Decimal('4.47')),
 ('AL', 17113602, 16933967, Decimal('1.06')),
 ('AR', 10762521, 10660058, Decimal('0.96')),
 ('KY', 19256394, 19113478, Decimal('0.75')),
 ('CO', 32978245, 32782247, Decimal('0.60')),
 ('SC', 18178677, 18105931, Decimal('0.40')),
 ('SD', 3899554, 3890392, Decimal('0.24')),
 ('MA', 42011647, 42237888, Decimal('-0.54')),
 ('AK', 3486955, 3525093, Decimal('-1.08')),
 ('ID', 8730670, 8847034, Decimal('-1.32')),
 ('NH', 7508751, 7675823, Decimal('-2.18')),
 ('WY', 3666825, 3756294, Decimal('-2.38')),
 ('MO', 29527698, 30330912, Decimal('-2.65')),
 ('MS', 9155488, 9456124, Decimal('-3.18')),
 ('VT', 3779545, 3919746, Decimal('-3.58')),
 ('ME', 6909414, 7190141, Decimal('-3.90')),
 ('WA', 41151390, 43451221, Decimal('-5.29')),
 ('NM', 7340034, 7800479, Decimal('-5.90')),
 ('AZ', 27552597, 29381343, Decimal('-6.22')),
 ('IA', 18042685, 19466908, Decimal('-7.32')),
 ('UT', 17524026, 18931103, Decimal('-7.43')),
 ('TX', 72876601, 78838400, Decimal('-7.56')),
 ('PA', 44791865, 49015763, Decimal('-8.62')),
 ('NV', 10525170, 11574976, Decimal('-9.07')),
 ('NE', 8333692, 9179047, Decimal('-9.21')),
 ('ND', 2181930, 2409123, Decimal('-9.43')),
 ('NJ', 45848755, 50701852, Decimal('-9.57')),
 ('TN', 19457120, 21544350, Decimal('-9.69')),
 ('CT', 21972583, 24454917, Decimal('-10.15')),
 ('WI', 32661683, 36467048, Decimal('-10.44')),
 ('CA', 162787836, 182181408, Decimal('-10.65')),
 ('OH', 82495138, 92402369, Decimal('-10.72')),
 ('KS', 13931862, 15673977, Decimal('-11.11')),
 ('NY', 106453546, 119810969, Decimal('-11.15')),
 ('VA', 37405643, 42300780, Decimal('-11.57')),
 ('WV', 5367596, 6081791, Decimal('-11.74')),
 ('IL', 72598213, 82438755, Decimal('-11.94')),
 ('DE', 3834672, 4357495, Decimal('-12.00')),
 ('OR', 21472271, 24511629, Decimal('-12.40')),
 ('MI', 49775454, 59037718, Decimal('-15.69')),
 ('IN', 35441059, 42199118, Decimal('-16.01')),
 ('FL', 73165352, 87730886, Decimal('-16.60')),
 ('MN', 24541150, 29534616, Decimal('-16.91')),
 ('HI', 4875365, 6019676, Decimal('-19.01')),
 ('MD', 27793912, 34474171, Decimal('-19.38')),
 ('RI', 5259143, 6612167, Decimal('-20.46')),
 ('NC', 33952977, 43111094, Decimal('-21.24')),
 ('PR', 193279, 257032, Decimal('-24.80')),
 ('GA', 28891017, 40922598, Decimal('-29.40')),
 ('OK', 13678542, 21171452, Decimal('-35.39'))]

#### Filtering an Aggregate Query Using HAVING

In [33]:
%%sql

SELECT pls14.stabr,
       sum(pls14.visits) AS visits_2014,
       sum(pls09.visits) AS visits_2009,
       round( (CAST(sum(pls14.visits) AS decimal(10,1)) - sum(pls09.visits)) /
                    sum(pls09.visits) * 100, 2 ) AS pct_change
FROM pls_fy2014_pupld14a pls14 JOIN pls_fy2009_pupld09a pls09
ON pls14.fscskey = pls09.fscskey
WHERE pls14.visits >= 0 AND pls09.visits >= 0
GROUP BY pls14.stabr
HAVING sum(pls14.visits) > 50000000
ORDER BY pct_change DESC;

 * postgresql://postgres:***@localhost:5432/analysis
6 rows affected.


[('TX', 72876601, 78838400, Decimal('-7.56')),
 ('CA', 162787836, 182181408, Decimal('-10.65')),
 ('OH', 82495138, 92402369, Decimal('-10.72')),
 ('NY', 106453546, 119810969, Decimal('-11.15')),
 ('IL', 72598213, 82438755, Decimal('-11.94')),
 ('FL', 73165352, 87730886, Decimal('-16.60'))]

* to filter the result of aggreagte functions we need to use **HAVING**
* using WHERE for filtering can't be used in aggregate functions where you have sum() functions, because aggregate functions work across rows
* the HAVING claise places conditions on groups created by aggregating 
* in this case ee've set pur query results to include only rows with a sum of visits on 2014 greater than 50 million

#### Try it Yourself

-- 1. We saw that library visits have declined in most places. But what is the
-- pattern in the use of technology in libraries? Both the 2014 and 2009 library
-- survey tables contain the columns gpterms (the number of internet-connected
-- computers used by the public) and pitusr (uses of public internet computers
-- per year). Modify the code in Listing 8-13 to calculate the percent change in
-- the sum of each column over time. Watch out for negative values!

In [42]:
%%sql

SELECT pls14.stabr,
        sum(pls14.gpterms) AS gpterms_2014,
        sum(pls09.gpterms) AS gpterms_2009,
        round( (CAST(sum(pls14.gpterms) AS decimal(10,1)) - sum(pls09.gpterms)) /
                    sum(pls09.gpterms) * 100, 2 ) AS pct_change_gpterms
FROM pls_fy2014_pupld14a pls14 JOIN pls_fy2009_pupld09a pls09
ON pls14.fscskey = pls09.fscskey
WHERE pls14.gpterms >= 0 AND pls09.gpterms >= 0 
GROUP BY pls14.stabr
ORDER BY pct_change_gpterms DESC;

 * postgresql://postgres:***@localhost:5432/analysis
53 rows affected.


[('GU', 547, 59, Decimal('827.12')),
 ('DC', 1000, 594, Decimal('68.35')),
 ('AK', 994, 618, Decimal('60.84')),
 ('DE', 772, 487, Decimal('58.52')),
 ('ID', 1792, 1151, Decimal('55.69')),
 ('CO', 6407, 4172, Decimal('53.57')),
 ('NE', 2942, 1955, Decimal('50.49')),
 ('ND', 841, 566, Decimal('48.59')),
 ('AZ', 5546, 3920, Decimal('41.48')),
 ('MT', 1262, 896, Decimal('40.85')),
 ('KY', 4832, 3531, Decimal('36.85')),
 ('IL', 15853, 11589, Decimal('36.79')),
 ('SC', 4203, 3082, Decimal('36.37')),
 ('ME', 1957, 1458, Decimal('34.22')),
 ('OK', 3061, 2324, Decimal('31.71')),
 ('GA', 8506, 6471, Decimal('31.45')),
 ('WI', 6234, 4879, Decimal('27.77')),
 ('WA', 6573, 5216, Decimal('26.02')),
 ('SD', 1099, 885, Decimal('24.18')),
 ('LA', 5359, 4330, Decimal('23.76')),
 ('OR', 3103, 2514, Decimal('23.43')),
 ('AR', 2501, 2028, Decimal('23.32')),
 ('TN', 5031, 4104, Decimal('22.59')),
 ('MS', 2744, 2240, Decimal('22.50')),
 ('MA', 6430, 5252, Decimal('22.43')),
 ('MD', 4858, 3982, Decimal('22.00')),
 ('NY', 19411, 15928, Decimal('21.87')),
 ('IA', 4719, 3873, Decimal('21.84')),
 ('NC', 7486, 6158, Decimal('21.57')),
 ('IN', 8759, 7218, Decimal('21.35')),
 ('VA', 6875, 5671, Decimal('21.23')),
 ('CA', 21416, 17789, Decimal('20.39')),
 ('UT', 2053, 1709, Decimal('20.13')),
 ('TX', 18849, 15847, Decimal('18.94')),
 ('MN', 5618, 4783, Decimal('17.46')),
 ('CT', 4355, 3714, Decimal('17.26')),
 ('MI', 11841, 10142, Decimal('16.75')),
 ('NH', 1473, 1262, Decimal('16.72')),
 ('AL', 5231, 4494, Decimal('16.40')),
 ('NJ', 8023, 6902, Decimal('16.24')),
 ('VT', 1089, 939, Decimal('15.97')),
 ('FL', 16282, 14047, Decimal('15.91')),
 ('WY', 868, 754, Decimal('15.12')),
 ('NM', 1619, 1416, Decimal('14.34')),
 ('KS', 3740, 3294, Decimal('13.54')),
 ('WV', 1450, 1280, Decimal('13.28')),
 ('OH', 12931, 11435, Decimal('13.08')),
 ('PA', 8050, 7226, Decimal('11.40')),
 ('RI', 1195, 1074, Decimal('11.27')),
 ('MO', 4930, 4436, Decimal('11.14')),
 ('NV', 1307, 1245, Decimal('4.98')),
 ('HI', 524, 537, Decimal('-2.42')),
 ('PR', 449, 518, Decimal('-13.32'))]

In [41]:
%%sql

SELECT pls14.stabr,
        sum(pls09.pitusr) AS pitusr_2009,
        sum(pls14.pitusr) AS pitusr_2014,
        round( (CAST(sum(pls14.pitusr) AS decimal(10,1)) - sum(pls09.pitusr)) /
                    sum(pls09.pitusr) * 100, 2 ) AS pct_change_pitusr
FROM pls_fy2014_pupld14a pls14 JOIN pls_fy2009_pupld09a pls09
ON pls14.fscskey = pls09.fscskey
WHERE pls14.pitusr >= 0  AND pls09.pitusr >= 0
GROUP BY pls14.stabr
ORDER BY pct_change_pitusr DESC;

 * postgresql://postgres:***@localhost:5432/analysis
53 rows affected.


[('DC', 140251, 1050623, Decimal('649.10')),
 ('GU', 19564, 39842, Decimal('103.65')),
 ('SD', 886453, 1472688, Decimal('66.13')),
 ('HI', 482309, 723305, Decimal('49.97')),
 ('DE', 451689, 622515, Decimal('37.82')),
 ('MT', 1346956, 1678243, Decimal('24.60')),
 ('AZ', 6971344, 8235693, Decimal('18.14')),
 ('MO', 5441405, 6375307, Decimal('17.16')),
 ('WA', 9281800, 10539089, Decimal('13.55')),
 ('ND', 519284, 577008, Decimal('11.12')),
 ('NM', 1995016, 2087226, Decimal('4.62')),
 ('AL', 4213655, 4328282, Decimal('2.72')),
 ('NY', 22322833, 22261108, Decimal('-0.28')),
 ('VA', 7309754, 7246441, Decimal('-0.87')),
 ('MS', 2551083, 2511033, Decimal('-1.57')),
 ('LA', 5442633, 5276086, Decimal('-3.06')),
 ('CO', 7672580, 7395748, Decimal('-3.61')),
 ('WY', 944682, 904784, Decimal('-4.22')),
 ('KY', 4753783, 4542269, Decimal('-4.45')),
 ('MD', 6834211, 6494653, Decimal('-4.97')),
 ('ID', 1986141, 1878131, Decimal('-5.44')),
 ('PA', 9091205, 8543924, Decimal('-6.02')),
 ('PR', 98945, 90351, Decimal('-8.69')),
 ('OH', 22036787, 20037428, Decimal('-9.07')),
 ('CA', 38631047, 34668471, Decimal('-10.26')),
 ('ME', 1344829, 1205962, Decimal('-10.33')),
 ('NJ', 11129337, 9799563, Decimal('-11.95')),
 ('TN', 6019840, 5258433, Decimal('-12.65')),
 ('GA', 15018105, 13003062, Decimal('-13.42')),
 ('NE', 2546524, 2130767, Decimal('-16.33')),
 ('UT', 3525463, 2924204, Decimal('-17.05')),
 ('FL', 22112567, 18340897, Decimal('-17.06')),
 ('MA', 8784952, 7279216, Decimal('-17.14')),
 ('TX', 21146858, 17431307, Decimal('-17.57')),
 ('IL', 17024802, 13884226, Decimal('-18.45')),
 ('OR', 4747789, 3867329, Decimal('-18.54')),
 ('IA', 4219992, 3421431, Decimal('-18.92')),
 ('SC', 5529774, 4481009, Decimal('-18.97')),
 ('RI', 1486973, 1201873, Decimal('-19.17')),
 ('NV', 3386816, 2736610, Decimal('-19.20')),
 ('CT', 5650168, 4465464, Decimal('-20.97')),
 ('NH', 1262597, 991583, Decimal('-21.46')),
 ('NC', 9752897, 7623395, Decimal('-21.83')),
 ('WI', 7798783, 6008759, Decimal('-22.95')),
 ('OK', 4711639, 3617868, Decimal('-23.21')),
 ('VT', 847682, 649906, Decimal('-23.33')),
 ('MI', 14519639, 10692717, Decimal('-26.36')),
 ('AK', 1061498, 771075, Decimal('-27.36')),
 ('MN', 7511316, 5438165, Decimal('-27.60')),
 ('IN', 10453312, 7226264, Decimal('-30.87')),
 ('WV', 1608829, 1075496, Decimal('-33.15')),
 ('AR', 3514858, 1922357, Decimal('-45.31')),
 ('KS', 6456844, 3165157, Decimal('-50.98'))]

-- 2. Both library survey tables contain a column called obereg, a two-digit
-- Bureau of Economic Analysis Code that classifies each library agency
-- according to a region of the United States, such as New England, Rocky
-- Mountains, and so on. Just as we calculated the percent change in visits
-- grouped by state, do the same to group percent changes in visits by US
-- regions using obereg. Consult the survey documentation to find the meaning
-- of each region code. For a bonus challenge, create a table with the obereg
-- code as the primary key and the region name as text, and join it to the
-- summary query to group by the region name rather than the code.


In [45]:
%%sql

SELECT pls14.obereg,
       sum(pls14.visits) AS visits_2014,
       sum(pls09.visits) AS visits_2009,
       round( (CAST(sum(pls14.visits) AS decimal(10,1)) - sum(pls09.visits)) /
                    sum(pls09.visits) * 100, 2 ) AS pct_change
FROM pls_fy2014_pupld14a pls14 JOIN pls_fy2009_pupld09a pls09
ON pls14.fscskey = pls09.fscskey
WHERE pls14.visits >= 0 AND pls09.visits >= 0
GROUP BY pls14.obereg
ORDER BY pct_change DESC;

 * postgresql://postgres:***@localhost:5432/analysis
9 rows affected.


[('07', 67482370, 68703182, Decimal('-1.78')),
 ('01', 87441083, 92090682, Decimal('-5.05')),
 ('09', 296872, 317795, Decimal('-6.58')),
 ('04', 100458571, 110484975, Decimal('-9.07')),
 ('08', 244298987, 271264003, Decimal('-9.94')),
 ('02', 232953540, 261305024, Decimal('-10.85')),
 ('06', 121447774, 137191674, Decimal('-11.48')),
 ('05', 289948497, 331552862, Decimal('-12.55')),
 ('03', 272971547, 312545008, Decimal('-12.66'))]

In [46]:
%%sql

CREATE TABLE obereg_codes (
    obereg varchar(2) CONSTRAINT obereg_key PRIMARY KEY,
    region varchar(50)
);

 * postgresql://postgres:***@localhost:5432/analysis
Done.


[]

In [47]:
%%sql

INSERT INTO obereg_codes
VALUES ('01', 'New England (CT ME MA NH RI VT)'),
       ('02', 'Mid East (DE DC MD NJ NY PA)'),
       ('03', 'Great Lakes (IL IN MI OH WI)'),
       ('04', 'Plains (IA KS MN MO NE ND SD)'),
       ('05', 'Southeast (AL AR FL GA KY LA MS NC SC TN VA WV)'),
       ('06', 'Soutwest (AZ NM OK TX)'),
       ('07', 'Rocky Mountains (CO ID MT UT WY)'),
       ('08', 'Far West (AK CA HI NV OR WA)'),
       ('09', 'Outlying Areas (AS GU MP PR VI)');

 * postgresql://postgres:***@localhost:5432/analysis
9 rows affected.


[]

In [49]:
%%sql

SELECT obereg_codes.region,
       sum(pls14.visits) AS visits_2014,
       sum(pls09.visits) AS visits_2009,
       round( (CAST(sum(pls14.visits) AS decimal(10,1)) - sum(pls09.visits)) /
                    sum(pls09.visits) * 100, 2 ) AS pct_change
FROM pls_fy2014_pupld14a pls14 JOIN pls_fy2009_pupld09a pls09
   ON pls14.fscskey = pls09.fscskey
JOIN obereg_codes
   ON pls14.obereg = obereg_codes.obereg
WHERE pls14.visits >= 0 AND pls09.visits >= 0
GROUP BY obereg_codes.region
ORDER BY pct_change DESC;

 * postgresql://postgres:***@localhost:5432/analysis
9 rows affected.


[('Rocky Mountains (CO ID MT UT WY)', 67482370, 68703182, Decimal('-1.78')),
 ('New England (CT ME MA NH RI VT)', 87441083, 92090682, Decimal('-5.05')),
 ('Outlying Areas (AS GU MP PR VI)', 296872, 317795, Decimal('-6.58')),
 ('Plains (IA KS MN MO NE ND SD)', 100458571, 110484975, Decimal('-9.07')),
 ('Far West (AK CA HI NV OR WA)', 244298987, 271264003, Decimal('-9.94')),
 ('Mid East (DE DC MD NJ NY PA)', 232953540, 261305024, Decimal('-10.85')),
 ('Soutwest (AZ NM OK TX)', 121447774, 137191674, Decimal('-11.48')),
 ('Southeast (AL AR FL GA KY LA MS NC SC TN VA WV)', 289948497, 331552862, Decimal('-12.55')),
 ('Great Lakes (IL IN MI OH WI)', 272971547, 312545008, Decimal('-12.66'))]

-- 3. Thinking back to the types of joins you learned in Chapter 6, which join
-- type will show you all the rows in both tables, including those without a
-- match? Write such a query and add an IS NULL filter in a WHERE clause to
-- show agencies not included in one or the other table.

In [50]:
%%sql

SELECT pls14.libname, pls14.city, pls14.stabr, pls14.statstru, pls14.c_admin, pls14.branlib,
       pls09.libname, pls09.city, pls09.stabr, pls09.statstru, pls09.c_admin, pls09.branlib
FROM pls_fy2014_pupld14a pls14 FULL OUTER JOIN pls_fy2009_pupld09a pls09
ON pls14.fscskey = pls09.fscskey
WHERE pls14.fscskey IS NULL OR pls09.fscskey IS NULL;

 * postgresql://postgres:***@localhost:5432/analysis
236 rows affected.


[('IPNATCHIAQ LIBRARY', 'DEERING', 'AK', '08', 'SO', 0, None, None, None, None, None, None),
 ('HOPE COMMUNITY LIBRARY', 'HOPE', 'AK', '00', 'SO', 0, None, None, None, None, None, None),
 ('KOYUK PUBLIC LIBRARY', 'KOYUK', 'AK', '00', 'SO', 0, None, None, None, None, None, None),
 ('OSCAR JOHNSON MEMORIAL LIBRARY', 'SILVERHILL', 'AL', '00', 'SO', 0, None, None, None, None, None, None),
 ('CLAY PUBLIC LIBRARY', 'PINSON', 'AL', '00', 'SO', 0, None, None, None, None, None, None),
 ('ECLECTIC PUBLIC LIBRARY', 'ECLECTIC', 'AL', '00', 'SO', 0, None, None, None, None, None, None),
 ('HOBSON CITY PUBLIC LIBRARY', 'HOBSON CITY', 'AL', '00', 'SO', 0, None, None, None, None, None, None),
 ('MUNFORD PUBLIC LIBRARY', 'MUNFORD', 'AL', '00', 'SO', 0, None, None, None, None, None, None),
 ('SOUTHSIDE PUBLIC LIBRARY', 'SOUTHSIDE', 'AL', '00', 'SO', 0, None, None, None, None, None, None),
 ('STEELE PUBLIC LIBRARY', 'STEELE', 'AL', '00', 'SO', 0, None, None, None, None, None, None),
 ('PINSON PUBLIC LIBRARY', 'PINSON', 'AL', '00', 'SO', 0, None, None, None, None, None, None),
 ('GLENCOE PUBLIC LIBRARY', 'GLENCOE', 'AL', '00', 'SO', 0, None, None, None, None, None, None),
 ('MAGNOLIA SPRINGS PUBLIC LIBRARY', 'MAGNOLIA SPRINGS', 'AL', '00', 'SO', 0, None, None, None, None, None, None),
 ('TOWN CREEK PUBLIC LIBRARY', 'TOWN CREEK', 'AL', '00', 'SO', 0, None, None, None, None, None, None),
 ('JEFFERSON COUNTY LIBRARY COOPERATIVE', 'BIRMINGHAM', 'AL', '00', 'SO', 0, None, None, None, None, None, None),
 ('MARENGO LIBRARY SYSTEM', 'DEMOPOLIS', 'AL', '00', 'SO', 0, None, None, None, None, None, None),
 ('PICKENS COUNTY COOPERATIVE LIBRARY', 'CARROLLTON', 'AL', '00', 'SO', 0, None, None, None, None, None, None),
 ('R.L. HADDEN MEMORIAL LIBRARY', 'HENAGAR', 'AL', '04', 'SO', 0, None, None, None, None, None, None),
 ('HEMPSTEAD COUNTY LIBRARY', 'HOPE', 'AR', '00', 'SO', 0, None, None, None, None, None, None),
 ('GENTRY PUBLIC LIBRARY', 'GENTRY', 'AR', '00', 'SO', 0, None, None, None, None, None, None),
 ('IVA JANE PEEK LIBRARY HQ', 'DECATUR', 'AR', '00', 'SO', 0, None, None, None, None, None, None),
 ('SILOAM SPRINGS PUBLIC LIBRARY HQ', 'SILOAM SPRINGS', 'AR', '00', 'SO', 0, None, None, None, None, None, None),
 ('SULPHUR SPRINGS PUBLIC LIBRARY HQ', 'SULPHUR SPRINGS', 'AR', '00', 'SO', 0, None, None, None, None, None, None),
 ('TRI-COUNTY REGIONAL LIBRARY SYSTEM', 'NASHVILLE', 'AR', '02', 'MA', 11, None, None, None, None, None, None),
 ('FELETI BARSTOW PUBLIC LIBRARY', 'PAGO PAGO', 'AS', '00', 'MO', 1, None, None, None, None, None, None),
 ('HOPI PUBLIC LIBRARY', 'KYKOTSMOVI', 'AZ', '00', 'MA', 0, None, None, None, None, None, None),
 ('AK-CHIN INDIAN COMMUNITY LIBRARY', 'MARICOPA', 'AZ', '00', 'SO', 0, None, None, None, None, None, None),
 ('SAN LUCY LIBRARY', 'GILA BEND', 'AZ', '00', 'SO', 0, None, None, None, None, None, None),
 ('SAN XAVIER LEARNING CENTER LIBRARY', 'TUCSON', 'AZ', '00', 'SO', 0, None, None, None, None, None, None),
 ('CAMARILLO PUBLIC LIBRARY', 'CAMARILLO', 'CA', '00', 'SO', 0, None, None, None, None, None, None),
 ('SANTA CLARITA PUBLIC LIBRARY', 'SANTA CLARITA', 'CA', '00', 'MO', 3, None, None, None, None, None, None),
 ('SIMI VALLEY PUBLIC LIBRARY', 'SIMI VALLEY', 'CA', '04', 'SO', 0, None, None, None, None, None, None),
 ('LINCOLN COUNTY BOOKMOBILE', 'LIMON', 'CO', '00', 'SO', 0, None, None, None, None, None, None),
 ('FLAGLER BEACH PUBLIC LIBRARY', 'FLAGLER BEACH', 'FL', '00', 'SO', 0, None, None, None, None, None, None),
 ('CITRUS SPRINGS MEMORIAL LIBRARY', 'CITRUS SPRINGS', 'FL', '00', 'SO', 0, None, None, None, None, None, None),
 ('PAL PUBLIC LIBRARY COOPERATIVE', 'PALATKA', 'FL', '00', 'MO', 23, None, None, None, None, None, None),
 ('SHALIMAR PUBLIC LIBRARY', 'SHALIMAR', 'FL', '00', 'SO', 0, None, None, None, None, None, None),
 ('TEMPLE TERRACE LIBRARY', 'TEMPLE TERRACE', 'FL', '04', 'SO', 0, None, None, None, None, None, None),
 ('BRUTON MEMORIAL LIBRARY', 'PLANT CITY', 'FL', '04', 'SO', 0, None, None, None, None, None, No